### ISSUE TITLE
Authentication Mechanism(s) for RideTeam

### ISSUE THREAD

**Issue by dustinmolieri** [RVTD/RideTeam#47]
I have a question on how we plan on performing authentication in RideTeam, or if we plan on having options (for each brokerage). IE:

AuthenticationOptions = { via = 'ad', dom='coic.org', grp='CERC' }
AuthenticationOptions = { via = 'mysql', auth='mysql.user.authentication_string', users='obss.obssuers' }
or for small brokerages
AuthenticationOptions = { via = 'csv', path='smb://medicaid/auth.csv' }

My favorite option is through the use of Active Directory where users are granted initial access by authenticating to a user in the domain and by belonging to a security group. I like resetting passwords and creating accounts for new staff in one place.

Once they've successfully identified themselves, then RideTeam's permissions begin to apply to the user (by username or SID).

ActiveDirectory

I only get about 4 to 6 hours a week to code as it is not my full time job, but yesterday I managed to complete an authentication system I intend to use for COIC's intranet in Node JS and I wanted to share how it works and see if it's something we'd like to adopt for rideteam. It took a total of about 30 hours to build, and the framework is fairly complete. I am still trying to break it (debug) but it seems solid so far.

The process is simple:

1) You visit the site (intranet, rideteam, whatever). You can browse areas that do not require authentication, such as the "About", "Support", Etc.. but once you hit a route that requires authentication, you must sign in or be signed in. Sign in is mandated by encapsulating the renderer calls in an authentication check function, ie:

router.get('/', function(req, res, next) {
  var options =
  {
    title: 'User Profile',
    supertitle: 'Profile',
  }
  authCheck(req,res,'cookie',(result) =>
  {
    if(result[0] != 201){
       const rProfile = pug.compileFile('views/sign_in_first.pug');
       var out = rProfile(options);
       res.send(out);
    } else {
       const rProfile = pug.compileFile('views/profile.pug');
       var out = rProfile(options);
       res.send(out);
    }
}); });

2) The server knows you must sign in because a cookie containing a token and another with your username is set when you sign in. If you are not signed in, the token cookie will not be set or it will be invalid. It will not return a 201.

3) In the authCheck method, before the page renderer displays the page or anything containing privileged data is responded to (in posts) the server checks for this token, authenticates it, passes a new token back, or fails and requests the user to sign in.

4) When a user signs in, the server creates a token using various meta data such as their IP address, the current UTC time, their username, and their password. This is encrypted using AES256-CBC (more on this in a moment). The token is then stored in a sessions table and also delivered back to the client where the client stores it as a cookie. the client's username is also stored as a cookie.

5) When the token is used to authenticate, first the server checks that the token matches the user's last token entered in the sessions database, then generates a new token and deletes the old one. The new token is passed back to the client and the client sets the new token in its token cookie.

**A new token is set every call to authCheck whether signing in for the first time, or authenticating via a token cookie.

Benefits to this System

6) AES256-CBC using CipherIV has a major security benefit in that if you encrypt the same contents twice, it will never appear the same twice. This, in addition the the changing UTC time ensure the token is unique with every authentication check. This is the cipher method used to create and encrypt tokens.

7) Because the token changes every ajax call, it expires quickly, and therefore cannot be placed on another computer (via a trojan or shared internet data folder, etc..).

8) Within the sessions database is also an expires segment which matches the cookie's expires segment. The server expires the cookie and does not rely on the client's browser to expire the cookie (there are add-ins that will do this so you do not get signed out automatically from certain sites)

9) We can remotely log someone out by simply removing their session from the database

10) The new expires segments are set every ajax call which allows us to inherently track inactivity. If a user is inactive for say... 15 minutes, they are automatically signed out at both the client and server.

How it works:

2 Server Scripts, 1 Client Script
authcheck.js (library) and interact.js (route) - Server
Since I am currently using PUG as the renderer, I embed all of the client JS inline with the layout.

The client JS is responsible A) For setting the token and username cookie, and B) for preventing the default behavior of the sign in form so it acts as an ajax call instead. When the client JS learns it is authenticated, it also sets certain content (such as, "You are signed in as ..." at the top of the page). It can be made to do more. It also renders the "sign out" option. I may move cookie setting back to the server end, but I had issues with doing this and opted for the client script to do this.

The authcheck JS contains the following methods:
Encrypt, Decrypt, GenTokenByForm, GenTokenByCookie, PerformAuth, which are all methods used in the primary method: authCheck(req,res,from,callback)

The key used for encryption is stored in the properties file along with the domain information where node-ActiveDirectory connects.

If you count the other lib.js (server side) which contains the methods used to interact with the properties file and other aspects, then technically it relies on 3 server js.

Thoughts?

Would you like me to create a project and post the code for you so you can see it yourself? See how you like it? Alternatively, I could do a short video.

Thanks,
- Dustin

**Comment by shemsargent-rvtd** [comment]
This would be useful for internal users. What about Providers, Facilities and Members, though? The current authentication method uses AES256 and does not depend on AD or MySQL authentication.

**Comment by dustinmolieri** [comment]
Good Point. I don't suppose we'd want providers, facilities, and members in the same AD as our staff.
